### PR TITLE
Rest API: rename navigation fallback classes from WP_ to Gutenberg_

### PR DESCRIPTION
**Pull request by ramonjd**
## What?

Renames `/compat/wordpress-6.3/class-wp-rest-navigation-fallback-controller.php` / `WP_REST_Navigation_Fallback_Controller` and associated classes/filenames to the `class-gutenberg-*` and `Gutenberg_*` conventions.

## Why?
The `WP_REST_Navigation_Fallback_Controller` class has been committed to core and therefore results in a naming conflict and unit test failures.

By renaming to `Gutenberg_*` we can avoid naming conflicts while still supporting 6.3 functionality in the plugin for users that won't upgrade to WordPress 6.3.

## How?
Renaming stuff.

## Testing Instructions
The unit tests should pass.

### CODE REVIEW

**Review by talldan — APPROVED** [review]
Thanks @ramonjd.

Looks like this is the approach other gutenebrg REST classes take, and tests are passing, so looks good to me!

File: packages/block-library/src/navigation/index.php [hunk 1]
```diff
@@ -846,7 +846,7 @@ function block_core_navigation_typographic_presets_backcompatibility( $parsed_bl
  */
 function block_core_navigation_parse_blocks_from_menu_items( $menu_items, $menu_items_by_parent_id ) {
 
-	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback_Gutenberg::parse_blocks_from_menu_items' );
+	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback::parse_blocks_from_menu_items' );
```
Comment (oandregal) [inline packages/block-library/src/navigation/index.php]: So, this deprecates the core class' method, not gutenberg's. Searched and that method is not available anywhere. I presume it was removed: where it was removed from? core's or gutenberg's class?

File: packages/block-library/src/navigation/index.php [hunk 2]
```diff
@@ -895,7 +895,7 @@ function block_core_navigation_parse_blocks_from_menu_items( $menu_items, $menu_
  */
 function block_core_navigation_get_classic_menu_fallback() {
 
-	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback_Gutenberg::get_classic_menu_fallback' );
+	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback::get_classic_menu_fallback' );
```
Comment (oandregal) [inline packages/block-library/src/navigation/index.php]: Same as https://github.com/WordPress/gutenberg/pull/51959/files#r1243377266

File: packages/block-library/src/navigation/index.php [hunk 3]
```diff
@@ -938,7 +938,7 @@ static function( $a, $b ) {
  */
 function block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_menu ) {
 
-	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback_Gutenberg::get_classic_menu_fallback_blocks' );
+	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback::get_classic_menu_fallback_blocks' );
```
Comment (oandregal) [inline packages/block-library/src/navigation/index.php]: Same as https://github.com/WordPress/gutenberg/pull/51959/files#r1243377266

File: packages/block-library/src/navigation/index.php (hunk 1 above)
Comment (ramonjd) [reply packages/block-library/src/navigation/index.php]: Thanks for checking @oandregal 

Here I thought the deprecation message should reflect the 6.3.0 class name.

But now I think of it, although `WP_Navigation_Fallback`  will be in 6.3, it isn't yet 😄 I'll update to refer to the Gutenberg class.

File: packages/block-library/src/navigation/index.php [hunk 4]
```diff
@@ -977,7 +977,7 @@ function block_core_navigation_get_classic_menu_fallback_blocks( $classic_nav_me
  */
 function block_core_navigation_maybe_use_classic_menu_fallback() {
 
-	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback_Gutenberg::create_classic_menu_fallback' );
+	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback::create_classic_menu_fallback' );
```
Comment (oandregal) [inline packages/block-library/src/navigation/index.php]: This method is present both in the core class and in gutenberg's. I'm not sure what's the purpose here, though I presume we'd want to deprecate the gutenberg one and only deprecate core when this code is backported?

File: packages/block-library/src/navigation/index.php (hunk 2 above)
Comment (ramonjd) [reply packages/block-library/src/navigation/index.php]: Same as above. Will rename

File: packages/block-library/src/navigation/index.php [hunk 5]
```diff
@@ -1020,7 +1020,7 @@ function block_core_navigation_maybe_use_classic_menu_fallback() {
  */
 function block_core_navigation_get_most_recently_published_navigation() {
 
-	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback_Gutenberg::get_most_recently_published_navigation' );
+	_deprecated_function( __FUNCTION__, '6.3.0', 'WP_Navigation_Fallback::get_most_recently_published_navigation' );
```
Comment (oandregal) [inline packages/block-library/src/navigation/index.php]: Same as https://github.com/WordPress/gutenberg/pull/51959/files#r1243380806

File: packages/block-library/src/navigation/index.php (hunk 4 above)
Comment (ramonjd) [reply packages/block-library/src/navigation/index.php]: You're right. I had a brain spasm. Fixed.

**Review by oandregal — APPROVED** [review]

**Comment by talldan** [comment]
I merged this with a failing performance test. That test checks out `trunk` and compares performance with this branch. When it checked out trunk it was re-encountering the issue that this PR fixes. I believe merging it should fix things.

**Comment by getdave** [comment]
Thanks for handling this. So for my reference:

- When authoring class like these in Plugin _always_ use `class-gutenberg-*` and `Gutenberg_*`.
- When backporting change all those instances to `class-wp-*` and `WP_*`.
- Put files in the directory for the release in which the features were backported to WP Core.

**Question**: is this documented? If not where could I contribute a docs update?

**Comment by ramonjd** [comment]
Hi @getdave 

No worries. You've got it spot on.

This document covers most of this, though not as succinctly as you've put it: https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md

Probably needs updating to include classes etc

**Comment by getdave** [comment]
[This](https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md#prefer-the-wp-prefix) seems to say the opposite of what I said above.

Perhaps it should say use `Gutenberg` until the point at when it's merged to Core?

**Comment by ramonjd** [comment]
Oops,  that document says:

> For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes.

So you were just following orders :) 

This probably needs to be cleared up. During release I don't think it's practical to juggle consistencies between plugin and Core where both have `wp_`, and renaming during backport isn't that hard.

🤷 

Sorry for the confusion. Despite that doc I think experience has led us to the `class-gutenberg-*` road.

I'll add a note to raise it with @noisysocks

**Comment by ramonjd** [comment]
I just cherry-picked this PR to the release/16.1 branch to get it included in the next release: 5a32952cd0

**Pull request merged**